File: multistore-wrong-country-tax-rate/node/audit-multistore-tax-rate.js
/**
 * Audit PrestaShop multistore orders for tax calculated at the wrong country's rate.
 *
 * Each shop in a multistore install has its own default country, but the tax
 * engine is supposed to resolve the rate from the invoice address's
 * id_country through PS_TAX_ADDRESS_TYPE and the TaxRulesGroup/TaxManager
 * classes. When the address is incomplete, when an order arrives through
 * pickup in store or the webservice without a full id_address_invoice, or
 * when a price context falls back to the shop's own country, the
 * TaxManager can silently use the shop's default country tax rule instead
 * of the customer's real one.
 *
 * This script reads a range of orders, recomputes the expected tax from
 * each order line's id_tax_rules_group and the invoice address's real
 * id_country, and compares it to the stored total_paid_tax_incl. It only
 * writes an audit report by default. A stored total tied to an invoice
 * must never be auto-corrected in place, so DRY_RUN=false only offers a
 * repair path for orders still in an editable, unpaid current_state, and
 * even then requires an explicit human confirmation before it writes
 * anything.
 *
 * Guide: https://www.allanninal.dev/prestashop/multistore-wrong-country-tax-rate/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/$/, "");
const WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const ID_ORDER_START = Number(process.env.ID_ORDER_START || 1);
const ID_ORDER_END = Number(process.env.ID_ORDER_END || 1);
const EDITABLE_STATE_NAMES = new Set(["awaiting payment", "awaiting check payment", "awaiting bank wire payment"]);

const EPSILON = 0.02;

function authHeader() {
  return "Basic " + Buffer.from(`${WS_KEY}:`).toString("base64");
}

export function computeExpectedTax(unitPriceTaxExcl, quantity, taxRatePct) {
  const expectedTaxExclTotal = Math.round(unitPriceTaxExcl * quantity * 100) / 100;
  const expectedTaxInclTotal = Math.round(expectedTaxExclTotal * (1 + taxRatePct / 100) * 100) / 100;
  return expectedTaxInclTotal;
}

export function selectApplicableTaxRate(orderCountryId, shopDefaultCountryId, taxRules) {
  for (const rule of taxRules) {
    if (Number(rule.id_country) === Number(orderCountryId)) return Number(rule.rate);
  }
  for (const rule of taxRules) {
    if (Number(rule.id_country) === Number(shopDefaultCountryId)) return Number(rule.rate);
  }
  return 0;
}

async function apiGet(path, params = {}) {
  const qs = new URLSearchParams({ ...params, output_format: "JSON" });
  const res = await fetch(`${PRESTASHOP_URL}/api/${path}?${qs}`, {
    headers: { Authorization: authHeader() },
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function apiPut(path, body) {
  const res = await fetch(`${PRESTASHOP_URL}/api/${path}?output_format=JSON`, {
    method: "PUT",
    headers: { Authorization: authHeader(), "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function apiPost(path, body) {
  const res = await fetch(`${PRESTASHOP_URL}/api/${path}?output_format=JSON`, {
    method: "POST",
    headers: { Authorization: authHeader(), "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function getOrder(idOrder) {
  const data = await apiGet(`orders/${idOrder}`, { display: "full" });
  return data.order || {};
}

async function getAddressCountry(idAddress) {
  const data = await apiGet(`addresses/${idAddress}`, { display: "full" });
  const raw = data.address?.id_country;
  return raw ? Number(raw) : null;
}

async function getOrderLines(idOrder) {
  const data = await apiGet("order_details", {
    "filter[id_order]": idOrder,
    display: "full",
  });
  return data.order_details || [];
}

async function getTaxRules(idTaxRulesGroup) {
  const data = await apiGet("tax_rules", {
    "filter[id_tax_rules_group]": idTaxRulesGroup,
    display: "full",
  });
  return data.tax_rules || [];
}

async function isEditableState(idOrderState) {
  const data = await apiGet(`order_states/${idOrderState}`, { display: "full" });
  let name = data.order_state?.name;
  if (name && typeof name === "object") name = Object.values(name)[0];
  return EDITABLE_STATE_NAMES.has(String(name || "").trim().toLowerCase());
}

async function scanOrder(idOrder) {
  const order = await getOrder(idOrder);
  const idShop = Number(order.id_shop || 0);
  const idAddressInvoice = Number(order.id_address_invoice || 0);
  const storedTotal = Number(order.total_paid_tax_incl || 0);
  const currentState = Number(order.current_state || 0);

  const orderCountryId = await getAddressCountry(idAddressInvoice);
  const shopData = await apiGet(`shops/${idShop}`, { display: "full" });
  const shopDefaultCountryId = Number(shopData.shop?.id_country || orderCountryId || 0);

  const lines = await getOrderLines(idOrder);
  let expectedTotal = 0;
  const lineFindings = [];
  for (const line of lines) {
    const idTaxRulesGroup = Number(line.id_tax_rules_group || 0);
    const unitPrice = Number(line.unit_price_tax_excl || 0);
    const quantity = Number(line.product_quantity || 0);
    const taxRules = await getTaxRules(idTaxRulesGroup);
    const rate = selectApplicableTaxRate(orderCountryId, shopDefaultCountryId, taxRules);
    const expectedLineTotal = computeExpectedTax(unitPrice, quantity, rate);
    expectedTotal += expectedLineTotal;
    lineFindings.push({
      id_order_detail: line.id,
      id_tax_rules_group: idTaxRulesGroup,
      unit_price_tax_excl: unitPrice,
      product_quantity: quantity,
      expected_rate: rate,
      expected_total_price_tax_incl: expectedLineTotal,
    });
  }

  if (Math.abs(storedTotal - expectedTotal) <= EPSILON) return null;

  return {
    id_order: idOrder,
    id_shop: idShop,
    id_address_invoice: idAddressInvoice,
    order_country_id: orderCountryId,
    shop_default_country_id: shopDefaultCountryId,
    current_state: currentState,
    stored_total_paid_tax_incl: storedTotal,
    expected_total_paid_tax_incl: Math.round(expectedTotal * 100) / 100,
    lines: lineFindings,
  };
}

async function applyCorrection(finding, confirmed) {
  if (!confirmed) {
    console.log(`Order ${finding.id_order}: correction available but not confirmed, skipping write.`);
    return;
  }
  if (!(await isEditableState(finding.current_state))) {
    console.warn(`Order ${finding.id_order}: current_state ${finding.current_state} is not editable, refusing to write.`);
    return;
  }

  for (const line of finding.lines) {
    await apiPut(`order_details/${line.id_order_detail}`, {
      total_price_tax_incl: line.expected_total_price_tax_incl,
      total_price_tax_excl: Math.round(line.unit_price_tax_excl * line.product_quantity * 100) / 100,
      unit_price_tax_incl: Math.round((line.expected_total_price_tax_incl / Math.max(line.product_quantity, 1)) * 100) / 100,
    });
  }

  const totalExcl = finding.lines.reduce((sum, l) => sum + l.unit_price_tax_excl * l.product_quantity, 0);
  await apiPut(`orders/${finding.id_order}`, {
    total_paid_tax_incl: finding.expected_total_paid_tax_incl,
    total_paid: finding.expected_total_paid_tax_incl,
    total_paid_tax_excl: Math.round(totalExcl * 100) / 100,
  });

  await apiPost("order_histories", {
    id_order: finding.id_order,
    id_order_state: finding.current_state,
  });
  console.log(`Order ${finding.id_order}: corrected to expected total ${finding.expected_total_paid_tax_incl.toFixed(2)}.`);
}

export async function run() {
  const findings = [];
  for (let idOrder = ID_ORDER_START; idOrder <= ID_ORDER_END; idOrder++) {
    const finding = await scanOrder(idOrder);
    if (finding) {
      findings.push(finding);
      console.warn(
        `Order ${finding.id_order} (shop ${finding.id_shop}): stored total_paid_tax_incl ${finding.stored_total_paid_tax_incl.toFixed(2)}, expected ${finding.expected_total_paid_tax_incl.toFixed(2)} for country ${finding.order_country_id}.`
      );
      if (!DRY_RUN) await applyCorrection(finding, false);
    }
  }
  console.log(`Done. ${findings.length} order(s) flagged for review.`);
  return findings;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: multistore-wrong-country-tax-rate/python/audit_multistore_tax_rate.py
"""Audit PrestaShop multistore orders for tax calculated at the wrong country's rate.

Each shop in a multistore install has its own default country, but the tax
engine is supposed to resolve the rate from the invoice address's
id_country through PS_TAX_ADDRESS_TYPE and the TaxRulesGroup/TaxManager
classes. When the address is incomplete, when an order arrives through
pickup in store or the webservice without a full id_address_invoice, or
when a price context falls back to the shop's own country, the TaxManager
can silently use the shop's default country tax rule instead of the
customer's real one.

This script reads a range of orders, recomputes the expected tax from each
order line's id_tax_rules_group and the invoice address's real id_country,
and compares it to the stored total_paid_tax_incl. It only writes an audit
report by default. A stored total tied to an invoice must never be
auto-corrected in place, so DRY_RUN=false only offers a repair path for
orders still in an editable, unpaid current_state, and even then requires
an explicit human confirmation before it writes anything.

Guide: https://www.allanninal.dev/prestashop/multistore-wrong-country-tax-rate/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("audit_multistore_tax_rate")

PRESTASHOP_URL = os.environ["PRESTASHOP_URL"].rstrip("/")
WS_KEY = os.environ["PRESTASHOP_WS_KEY"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
ID_ORDER_START = int(os.environ.get("ID_ORDER_START", "1"))
ID_ORDER_END = int(os.environ.get("ID_ORDER_END", "1"))
EDITABLE_STATE_NAMES = {"awaiting payment", "awaiting check payment", "awaiting bank wire payment"}

EPSILON = 0.02


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(
        f"{PRESTASHOP_URL}/api/{path}",
        params=params,
        auth=(WS_KEY, ""),
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def api_put(path, body):
    r = requests.put(
        f"{PRESTASHOP_URL}/api/{path}",
        params={"output_format": "JSON"},
        json=body,
        auth=(WS_KEY, ""),
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def api_post(path, body):
    r = requests.post(
        f"{PRESTASHOP_URL}/api/{path}",
        params={"output_format": "JSON"},
        json=body,
        auth=(WS_KEY, ""),
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def compute_expected_tax(unit_price_tax_excl, quantity, tax_rate_pct):
    """Pure math: expected tax inclusive total for one order line."""
    expected_tax_excl_total = round(unit_price_tax_excl * quantity, 2)
    expected_tax_incl_total = round(expected_tax_excl_total * (1 + tax_rate_pct / 100), 2)
    return expected_tax_incl_total


def select_applicable_tax_rate(order_country_id, shop_default_country_id, tax_rules):
    """
    tax_rules: list of {"id_country": int, "rate": float} for one id_tax_rules_group.
    Must select the rule for order_country_id (the invoice address country), and must
    NOT fall back to shop_default_country_id when a matching rule for order_country_id
    exists and the two ids differ. Pure decision logic, no I/O.
    """
    for rule in tax_rules:
        if int(rule["id_country"]) == int(order_country_id):
            return float(rule["rate"])
    for rule in tax_rules:
        if int(rule["id_country"]) == int(shop_default_country_id):
            return float(rule["rate"])
    return 0.0


def get_order(id_order):
    data = api_get(f"orders/{id_order}", {"display": "full"})
    return data.get("order") or {}


def get_address_country(id_address):
    data = api_get(f"addresses/{id_address}", {"display": "full"})
    address = data.get("address") or {}
    return int(address.get("id_country") or 0) or None


def get_order_lines(id_order):
    data = api_get("order_details", {
        "filter[id_order]": id_order,
        "display": "full",
    })
    return data.get("order_details") or []


def get_tax_rules(id_tax_rules_group):
    data = api_get("tax_rules", {
        "filter[id_tax_rules_group]": id_tax_rules_group,
        "display": "full",
    })
    return data.get("tax_rules") or []


def is_editable_state(id_order_state):
    data = api_get(f"order_states/{id_order_state}", {"display": "full"})
    state = data.get("order_state") or {}
    name = state.get("name")
    if isinstance(name, dict):
        name = next(iter(name.values()), "")
    return str(name or "").strip().lower() in EDITABLE_STATE_NAMES


def scan_order(id_order):
    order = get_order(id_order)
    id_shop = int(order.get("id_shop") or 0)
    id_address_invoice = int(order.get("id_address_invoice") or 0)
    stored_total = float(order.get("total_paid_tax_incl") or 0)
    current_state = int(order.get("current_state") or 0)

    order_country_id = get_address_country(id_address_invoice)
    shop = api_get(f"shops/{id_shop}", {"display": "full"}).get("shop") or {}
    shop_default_country_id = int(shop.get("id_country") or order_country_id or 0)

    lines = get_order_lines(id_order)
    expected_total = 0.0
    line_findings = []
    for line in lines:
        id_tax_rules_group = int(line.get("id_tax_rules_group") or 0)
        unit_price = float(line.get("unit_price_tax_excl") or 0)
        quantity = int(line.get("product_quantity") or 0)
        tax_rules = get_tax_rules(id_tax_rules_group)
        rate = select_applicable_tax_rate(order_country_id, shop_default_country_id, tax_rules)
        expected_line_total = compute_expected_tax(unit_price, quantity, rate)
        expected_total += expected_line_total
        line_findings.append({
            "id_order_detail": line.get("id"),
            "id_tax_rules_group": id_tax_rules_group,
            "unit_price_tax_excl": unit_price,
            "product_quantity": quantity,
            "expected_rate": rate,
            "expected_total_price_tax_incl": expected_line_total,
        })

    if abs(stored_total - expected_total) <= EPSILON:
        return None

    return {
        "id_order": id_order,
        "id_shop": id_shop,
        "id_address_invoice": id_address_invoice,
        "order_country_id": order_country_id,
        "shop_default_country_id": shop_default_country_id,
        "current_state": current_state,
        "stored_total_paid_tax_incl": stored_total,
        "expected_total_paid_tax_incl": round(expected_total, 2),
        "lines": line_findings,
    }


def apply_correction(finding, confirmed):
    if not confirmed:
        log.info("Order %s: correction available but not confirmed, skipping write.", finding["id_order"])
        return
    if not is_editable_state(finding["current_state"]):
        log.warning("Order %s: current_state %s is not editable, refusing to write.",
                    finding["id_order"], finding["current_state"])
        return

    for line in finding["lines"]:
        api_put(f"order_details/{line['id_order_detail']}", {
            "total_price_tax_incl": line["expected_total_price_tax_incl"],
            "total_price_tax_excl": round(line["unit_price_tax_excl"] * line["product_quantity"], 2),
            "unit_price_tax_incl": round(line["expected_total_price_tax_incl"] / max(line["product_quantity"], 1), 2),
        })

    api_put(f"orders/{finding['id_order']}", {
        "total_paid_tax_incl": finding["expected_total_paid_tax_incl"],
        "total_paid": finding["expected_total_paid_tax_incl"],
        "total_paid_tax_excl": round(sum(l["unit_price_tax_excl"] * l["product_quantity"] for l in finding["lines"]), 2),
    })

    api_post("order_histories", {
        "id_order": finding["id_order"],
        "id_order_state": finding["current_state"],
    })
    log.info("Order %s: corrected to expected total %.2f.", finding["id_order"], finding["expected_total_paid_tax_incl"])


def run():
    findings = []
    for id_order in range(ID_ORDER_START, ID_ORDER_END + 1):
        finding = scan_order(id_order)
        if finding:
            findings.append(finding)
            log.warning(
                "Order %s (shop %s): stored total_paid_tax_incl %.2f, expected %.2f for country %s.",
                finding["id_order"], finding["id_shop"],
                finding["stored_total_paid_tax_incl"], finding["expected_total_paid_tax_incl"],
                finding["order_country_id"],
            )
            if not DRY_RUN:
                apply_correction(finding, confirmed=False)
    log.info("Done. %d order(s) flagged for review.", len(findings))
    return findings


if __name__ == "__main__":
    run()
